From 6a378633409d68de63404f288d34974a3f5997cc Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 31 Dec 2021 10:51:53 +0100
Subject: [PATCH 53/64] Revert "block, bfq: update pos_root for idle bfq_queue
 in bfq_bfqq_move()"

This reverts commit 5bd8eadf6287679759b8e43d0c90bfe0d5a448e3.

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-cgroup.c | 14 +++++---------
 1 file changed, 5 insertions(+), 9 deletions(-)

diff --git a/block/bfq-cgroup.c b/block/bfq-cgroup.c
index d41fd6f..4b41754 100644
--- a/block/bfq-cgroup.c
+++ b/block/bfq-cgroup.c
@@ -678,6 +678,7 @@ void bfq_bfqq_move(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 	else if (entity->on_st_or_in_serv)
 		bfq_put_idle_entity(bfq_entity_service_tree(entity), entity);
 	hlist_del(&bfqq->children_node);
+	bfqg_and_blkg_put(old_parent);
 
 	if (entity->parent &&
 	    entity->parent->last_bfqq_created == bfqq)
@@ -691,16 +692,11 @@ void bfq_bfqq_move(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 	bfqg_and_blkg_get(bfqg);
 	hlist_add_head(&bfqq->children_node, &bfqg->children);
 
-	/*
-	 * Don't leave the pos_root to old bfqg, since the ref to old bfqg will
-	 * be released and the bfqg might be freed.
-	 */
-	if (unlikely(!bfqd->nonrot_with_queueing))
-		bfq_pos_tree_add_move(bfqd, bfqq);
-	bfqg_and_blkg_put(old_parent);
-
-	if (bfq_bfqq_busy(bfqq))
+	if (bfq_bfqq_busy(bfqq)) {
+		if (unlikely(!bfqd->nonrot_with_queueing))
+			bfq_pos_tree_add_move(bfqd, bfqq);
 		bfq_activate_bfqq(bfqd, bfqq);
+	}
 
 	if (!bfqd->in_service_queue && !bfqd->rq_in_driver)
 		bfq_schedule_dispatch(bfqd);
-- 
2.34.1.75.gabe6bb3905

